spec/zipkin_spec.lua

```diff
@@ -168,22 +173,35 @@ describe("integration tests with zipkin server [#" .. strategy .. "]", function(
     local request_tags = request_span.tags
     assert.truthy(request_tags["kong.node.id"]:match("^[%x-]+$"))
     request_tags["kong.node.id"] = nil
+    local consumer_port = request_tags["kong.consumer.port"]
+    assert_is_integer(tonumber(consumer_port))
     assert.same({
       ["http.method"] = "GET",
       ["http.path"] = "/",
       ["http.status_code"] = "200", -- found (matches server status)
+      ["kong.route"] = route.id,
+      ["kong.service"] = route.service.id,
+      ["kong.consumer.port"] = consumer_port,
```

Comment:
if this is a server span, and the consumer is the client, then this ip/port can move to remoteEndpoint vs being tags.

I'm not sure the concept of consumer or if this is a server span, so if this isn't the case, ignore

Reply:
> if this is a server span, and the consumer is the client, then this ip/port can move to remoteEndpoint vs being tags.

I think that is what we had before. Let me see if I have understood correctly.

On an API gateway scenario (the most common one for Kong at the moment) Kong acts as a proxy between a consumer (downstream) and a Service (upstream).

```
Request:
Consumer   --->   Kong  --->  Service

Response:
Consumer   <---   Kong  <---  Service
```

The *root span* generated by the zipkin plugin (kind = server) starts when Kong receives the consumer request, and ends when kong sends the response back to the consumer. So this span's remoteEndpoint's entity should point to the consumer.

Kong generates more spans while it communicates with the Service (at least 2, kind = client) and the Consumer isn't involved there. You are saying that the remoteEndpoint entity on those should be the Service.

On the other hand, `localEndpoint` should always be Kong.

Did I understand this correctly?

Reply:
I agree that only the direct remote span should add the remoteEndpoint. In other words, it doesn't propagate to span data in children and so forth. I think what was confusing me was presence of the consumer tags.

I guess you are copying these to other spans for some reason. At least in zipkin this isn't required for search because search is trace scoped. Copying tags to children (about the consumer/client in this case) would add index load and increase the data size. If were me, if adding additional tags about the consumer/client I'd put in a comment as to why to prevent questions like this later.

Reply:
Hi, I apologize, but I have re-read your last comment and I still don't understand. Could you please spell out what you think `remoteEndpoint` should be on each of the spans? Right now all spans will "point" to the Service, and all `localEndpoints` point to Kong. Should I change anything (i.e. make root span's remoteEndpoint be the consumer) or not?

> I think what was confusing me was presence of the consumer tags.

The way I see it is that for Kong there isn't a single "remote entity", there are at least two (the consumer and the Service). I can put one or the other as a `remoteEntity`, but I thought it would be useful to add the other one somewhere. I don't know which one is more convenient or appropriate.

Reply:
lets focus this on remoteEndpoint

a span of kind server, its remoteEndpoint is the caller which in RPC terms is the upstream client
similarly a span of kind server, its remoteEndpoint is who it is calling, in RPC terms the downstream server

as a span is either client or server and not both it should be unambiguous what to place into the remoteEndpoint without further need to clarify with tags. if there is an intermediate span between the server and the client on the same host in the same request it has no remoteendpoint and no need to carry any RPC related properties.

the additional terminology of consumer and service seem to be directionally oriented to me.. like labeling the two sides of the proxy and be the remoteEndpoint.serviceName of a client or server span depending on the direction. I dont understand why there would need to be other tags about this but I have sortof said all I think I can.

Reply:
not sure it helps but in zipkin the endpoint.serviceName is the name of a node independent of direction. the labels of the node called or being called go in this field when known. that's where I think these consumer or service names should go, but I honestly dont know the subtleties which are making this confusing

Reply:
I think what Adrian means is:

**For Consumer   --->   Kong  --->  Service**
- for `Consumer   --->   Kong` this `SERVER` span should have the `consumer IP` at `remoteEndpoint`
- for `Kong  --->  Service` this `CLIENT` span should have the `service IP` at `remoteEndpoint`

In either case `consumer IP` should not be a tag. Any other span that does not involve a remote call is a local span and that does not need a `remoteEndpoint`.

Are there any other spans that involve remote calls?

Reply:
> a span of kind server, its remoteEndpoint is the caller 

Ok, I will make remoteEndpoint point to the consumer on the root span, and to the kong service on the internal spans. And not use custom kong tags for this.